driver-core/src/test/java/com/datastax/driver/core/PagingStateTest.java

```diff
@@ -0,0 +1,109 @@
+
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNull;
+
+import com.datastax.driver.core.exceptions.InvalidPagingStateException;
+
+
+public class PagingStateTest extends CCMBridge.PerClassSingleNodeCluster {
+
+    @Override
+    protected Collection<String> getTableDefinitions() {
+        return Collections.singletonList("CREATE TABLE test (k text, v int, PRIMARY KEY (k, v))");
+    }
+
+    @Test(groups = "short")
+    public void changeInPagingStateFailsVerification (){
+
+        String key = "paging_test";
+        for (int i = 0; i < 100; i++)
+            session.execute(String.format("INSERT INTO test (k, v) VALUES ('%s', %d)", key, i));
+
+        SimpleStatement st = new SimpleStatement(String.format("SELECT v FROM test WHERE k='%s'", key));
+        ResultSet result = session.execute(st.setFetchSize(20));
+
+        Iterator<Row> rowIt = result.iterator();
+        for (int i = 0; i < 30; i++) {
+            int value = rowIt.next().getInt("v");
+        }
+        PagingState savedPagingState = result.getExecutionInfo().getPagingState();
+        String savedPagingStateString = savedPagingState.toString();
+
+        // corrupting the paging state
+        savedPagingState.getRawState().putInt(42);
+        ByteBuffer pagingstate = savedPagingState.getRawState().duplicate();
+
+        st = new SimpleStatement(String.format("SELECT v FROM test WHERE k='%s'", key));
+        try {
+            st.setFetchSize(20).setPagingState(savedPagingState);
+        }
+        catch (InvalidPagingStateException e) {
+            System.out.println(e.getMessage());
+        }
+        finally{
+            assertNull(st.getPagingState());
+        }
+
+        st = new SimpleStatement(String.format("SELECT v FROM test WHERE k='%s'", "paging"));
+        try {
+            st.setFetchSize(20).setPagingState(savedPagingStateString);
+        }
+        catch (InvalidPagingStateException e) {
+            System.out.println(e.getMessage());
+        }
+        finally {
+            assertNull(st.getPagingState());
+        }
+    }
+
+    @Test(groups = "short")
+    public void getAndUseThePagingState() {
+
+        String key = "paging_test";
+
+        SimpleStatement st = new SimpleStatement(String.format("SELECT v FROM test WHERE k='%s'", key));
+        ResultSet result = session.execute(st.setFetchSize(20));
+
+        Iterator<Row> rowIt = result.iterator();
+        for (int i = 0; i < 30; i++) {
+            int value = rowIt.next().getInt("v");
+        }
+
+        String savedPagingStateString = result.getExecutionInfo().getPagingState().toString();
+
+        st = new SimpleStatement(String.format("SELECT v FROM test WHERE k='%s'", key));
+        result = session.execute(st.setFetchSize(20).setPagingState(savedPagingStateString));
+
+        //We have the result starting from the next page we stopped
+        assertEquals(result.one().getInt("v"), 40);
```

Comment:
I'm splitting hairs, but fetch size is only an indication of the *desired* fetch size. Cassandra is not required to respect it exactly, it can return less rows (see section 8 in the native protocol spec).
What you could do instead is:
* run the initial query (no need to iterate over the results)
* save `rs.getAvailableWithoutFetching()` (the size of the first page that just came back) to a variable
* run the query a second time with the paging state from the previous result
* ensure that the first result for this query is at the index matching the previously saved variable